frb_rust/src/ffi/web.rs

```diff
@@ -0,0 +1,207 @@
+    Vec<f32> => js_sys::Float32Array
+    Vec<f64> => js_sys::Float64Array
+    ZeroCopyBuffer<Vec<i8>> => js_sys::Int8Array
+    ZeroCopyBuffer<Vec<u8>> => js_sys::Uint8Array
+    ZeroCopyBuffer<Vec<i16>> => js_sys::Int16Array
+    ZeroCopyBuffer<Vec<u16>> => js_sys::Uint16Array
+    ZeroCopyBuffer<Vec<i32>> => js_sys::Int32Array
+    ZeroCopyBuffer<Vec<u32>> => js_sys::Uint32Array
+    ZeroCopyBuffer<Vec<f32>> => js_sys::Float32Array
+    ZeroCopyBuffer<Vec<f64>> => js_sys::Float64Array
+}
+
+impl<T: IntoDartExceptPrimitive> IntoDart for Vec<T> {
+    #[inline]
+    fn into_dart(self) -> DartCObject {
+        Array::from_iter(self.into_iter().map(IntoDart::into_dart)).into()
+    }
+}
+
+impl<T: IntoDart> IntoDart for Option<T> {
+    #[inline]
+    fn into_dart(self) -> DartCObject {
+        self.map(T::into_dart).unwrap_or_else(JsValue::null)
+    }
+}
+
+impl<T: IntoDart + Clone> IntoDart for &T {
+    #[inline]
+    fn into_dart(self) -> DartCObject {
+        self.clone().into_dart()
+    }
+}
+
+#[derive(Clone)]
+pub struct Isolate {
+    port: MessagePort,
+}
+
+impl Isolate {
+    pub fn new(port: MessagePort) -> Self {
+        Self { port }
+    }
+    pub fn post(&self, msg: impl IntoDart) -> bool {
+        self.port
+            .post_message(&msg.into_dart())
+            .map_err(|err| {
+                crate::console_error!("post: {:?}", err);
+                err
+            })
+            .is_ok()
+    }
+}
+
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(js_namespace = console)]
+    pub fn log(msg: &str);
+    #[wasm_bindgen(js_namespace = console)]
+    pub fn error(msg: &str);
+    // fn eval(script: &str);
+}
+
+type RawClosure<T> = Box<dyn FnOnce(&[T]) + Send + 'static>;
+
+pub struct TransferClosure<T> {
+    transfer: Vec<T>,
+    closure: RawClosure<T>,
+}
+
+pub struct TransferClosurePayload<T> {
+    func: RawClosure<T>,
+}
+
+impl TransferClosure<JsValue> {
+    pub fn new(transfer: Vec<JsValue>, closure: impl FnOnce(&[JsValue]) + Send + 'static) -> Self {
+        let closure = Box::new(closure);
+        Self { transfer, closure }
+    }
+    /// Posts a <code>[*mut [TransferClosurePayload], ...[JsValue]]</code> message to this worker.
+    ///
+    /// The worker's `onmessage` should run the corresponding [`receive_transfer_closure`]
+    /// to receive the message.
+    pub fn apply(self, worker: &Worker) -> Result<(), JsValue> {
+        let transfer = Array::from_iter(self.transfer);
+        // The worker is responsible for cleaning up the leak here.
+        let payload = Box::into_raw(Box::new(TransferClosurePayload { func: self.closure }));
+        let data = Array::from(&transfer);
+        data.unshift(&JsValue::from(payload as i32));
+        worker
+            .post_message_with_transfer(&data, &transfer)
+            .map_err(|err| {
+                // post message failed, ownership remains with us.
+                drop(unsafe { Box::from_raw(payload) });
```

Comment:
Is it possible we add some quick unit tests here - the "unsafe" looks a bit, well, unsafe...

Reply:
This is part of the original pool implementation from wasm_bindgen so I have good reason to believe it is correct. Regardless, I can try to write something to test this.

Reply:
I see. So can we use the code inside wasm_bindgen, instead of copy-and-paste here?

If not, at least we should make a comment telling readers where this code comes from. Indeed, we may *extract everything copying from wasm_bindgen into a separate .rs file with names reflecting the fact*. 

For example, `utils_copied_from_wasm_bindgen.rs` or sth like that.

By doing so, it is easier to believe the code is right.